File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
import heapq
def solution(numbers):
    answer = [-1]*len(numbers)
    stack = []
    for i in range(0, len(numbers)):
        while stack and stack[-1][1] < numbers[i]:
            id, now= stack.pop()
            answer[id] = numbers[i]
        stack.append((i, numbers[i]))
    
    while stack:
        id,now = stack.pop()
        answer[id] = -1
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;
class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        
        answer[numbers.length-1] = -1;
        Stack<Integer> stack = new Stack<>();
        for(int i=numbers.length-1; i>-1;i--){
            while(!stack.isEmpty() && stack.peek() <= numbers[i]){
                stack.pop();
            }
            answer[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(numbers[i]);
        }
        
        return answer;
    }
}